pkg/proxy/metaproxier/meta_proxier.go

```diff
@@ -79,11 +103,88 @@ func (proxier *metaProxier) OnServiceUpdate(oldService, service *v1.Service) {
 	if utilproxy.ShouldSkipService(service) {
 		return
 	}
-	// IPFamily is immutable, hence we only need to check on the new service
-	if utilnet.IsIPv6String(service.Spec.ClusterIP) {
-		proxier.ipv6Proxier.OnServiceUpdate(oldService, service)
-	} else {
-		proxier.ipv4Proxier.OnServiceUpdate(oldService, service)
+	// case zero: this allows skew between new proxy and old apiserver
+	if len(service.Spec.IPFamilies) == 0 {
+		actual := proxier.getProxyByClusterIP(oldService)
+		actual.OnServiceUpdate(oldService, service)
+		return
+	}
+
+	// case one: something has changed, but not families
+	// call update on all families the service carries.
+	if len(oldService.Spec.IPFamilies) == len(service.Spec.IPFamilies) {
+		for _, ipFamily := range service.Spec.IPFamilies {
+			actual := proxier.getProxierByIPFamily(ipFamily)
+			actual.OnServiceUpdate(oldService, service)
+		}
+
+		klog.V(4).Infof("service %s/%s has been updated but no ip family change detected", service.Namespace, service.Name)
+		return
+	}
+	// while apiserver does not allow changing primary ipfamily
+	// we use the below approach to stay on the safe side.
+	// note: in all cases, we check all families just
+	// in case the service moved from ExternalName => ClusterIP
+	// or the other way around
```

Comment:
this is not possible , an Externalname will be filtered at the beginning of this method by 
```
	if utilproxy.ShouldSkipService(service) {
		return
	}
```